sdk/src/trace/batch_span_processor.cc

```diff
@@ -0,0 +1,285 @@
+
+#include "opentelemetry/sdk/trace/batch_span_processor.h"
+#include "src/common/circular_buffer.h"
+
+#include <vector>
+using opentelemetry::sdk::common::AtomicUniquePtr;
+using opentelemetry::sdk::common::CircularBuffer;
+using opentelemetry::sdk::common::CircularBufferRange;
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace trace
+{
+BatchSpanProcessor::BatchSpanProcessor(std::unique_ptr<SpanExporter> &&exporter,
+                                       const size_t max_queue_size,
+                                       const std::chrono::milliseconds schedule_delay_millis,
+                                       const size_t max_export_batch_size)
+    : exporter_(std::move(exporter)),
+      max_queue_size_(max_queue_size),
+      schedule_delay_millis_(schedule_delay_millis),
+      max_export_batch_size_(max_export_batch_size)
+{
+  buffer_ =
+      std::unique_ptr<CircularBuffer<Recordable>>(new CircularBuffer<Recordable>(max_queue_size_));
+
+  // Start the background worker thread
+  worker_thread_ =
+      std::unique_ptr<std::thread>(new std::thread(&BatchSpanProcessor::DoBackgroundWork, this));
+}
+
+std::unique_ptr<Recordable> BatchSpanProcessor::MakeRecordable() noexcept
+{
+  return exporter_->MakeRecordable();
+}
+
+void BatchSpanProcessor::OnStart(Recordable &span) noexcept
+{
+  // no-op
+  (void)span;
+}
+
+void BatchSpanProcessor::OnEnd(std::unique_ptr<Recordable> &&span) noexcept
+{
+  if (is_shutdown_ == true)
+  {
+    // TODO: log in glog
+    return;
+  }
+
+  std::unique_lock<std::mutex> lk(cv_m_);
```

Comment:
This is because we make a copy of the `buffer_` in the worker thread and then export that copy buffer. While copying, we ensure with this lock that the `buffer_` is not manipulated.

Reply:
I think you don't need to make a copy of the buffer. You can consume while producers are adding spans.

Reply:
The purpose of making a copy of the buffer was to record the exact state of the buffer at a given point in time, and subsequently empty out the buffer for other spans.

Let's say ForceFlush was invoked, which in turned required that the entire buffer be flushed out at that very instance. In that case, if we don't record/duplicate the exact state of the buffer when the ForceFlush method was invoked, it's possible that other producers keep adding spans to the buffer while the buffer is being consumed in the Export method. This raises the concern that some spans, which shouldn't have been dropped, get dropped.

Reply:
> Let's say ForceFlush was invoked, which in turned required that the entire buffer be flushed out at that very instance.

With the current solution, there's some time elapsing between the moment the user calls `ForceFlush` and copying the buffer (the time you wait until `is_force_flush_` is set to `true`). Spans might be added to the buffer during this time, thus resulting in spans being dropped from the buffer (when it is full).

I think we can't give that strong of a guarantee. And even if we could it is not of much use to the user, as they cannot exactly know _what_ spans are in the buffer at the time of the `ForceFlush` invocation.

Also, copying the whole buffer and adding locks around it is pretty expensive, considering that this happens for each iteration and that the users can (and will) use large buffers.

Reply:
Also, it's important that we avoid any locking in `OnEnd`, as this is invoked on the user thread.